File: src/adultgen/api/schemas/subscriptions.py
"""Subscription API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class SubscriptionPlanResponse(BaseModel):
    """Subscription plan exposed to the website."""

    code: str
    title: str
    credits_per_period: int
    amount_minor: int
    amount_major: str
    currency: str
    period_days: int
    rollover_policy: str
    description: str
    is_popular: bool


class SubscriptionPlanListResponse(BaseModel):
    """List of enabled subscription plans."""

    items: list[SubscriptionPlanResponse]


class ActivateSubscriptionRequest(BaseModel):
    """Activate an MVP subscription plan."""

    plan_code: str
    provider: str | None = None
    provider_subscription_id: str | None = None


class SubscriptionResponse(BaseModel):
    """User subscription state."""

    id: uuid.UUID
    plan_code: str
    status: str
    provider: str | None
    provider_subscription_id: str | None
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool
    cancelled_at: datetime | None
    last_granted_at: datetime | None


class SubscriptionActivationResponse(BaseModel):
    """Activation response with grant info."""

    subscription: SubscriptionResponse
    plan: SubscriptionPlanResponse
    granted_now: bool
    grant_id: uuid.UUID | None
    credits_granted: int

File: apps/web_app/src/subscriptions.ts
export type SubscriptionPlan = {
  code: string;
  title: string;
  credits_per_period: number;
  amount_minor: number;
  amount_major: string;
  currency: string;
  period_days: number;
  rollover_policy: string;
  description: string;
  is_popular: boolean;
};

export type Subscription = {
  id: string;
  plan_code: string;
  status: string;
  provider: string | null;
  provider_subscription_id: string | null;
  current_period_start: string;
  current_period_end: string;
  cancel_at_period_end: boolean;
  cancelled_at: string | null;
  last_granted_at: string | null;
};

export type SubscriptionActivationResponse = {
  subscription: Subscription;
  plan: SubscriptionPlan;
  granted_now: boolean;
  grant_id: string | null;
  credits_granted: number;
};

const CORE_API_URL = import.meta.env.VITE_CORE_API_URL || '/api';

export async function fetchSubscriptionPlans(): Promise<{ items: SubscriptionPlan[] }> {
  return request<{ items: SubscriptionPlan[] }>('/subscriptions/plans');
}

export async function fetchMySubscription(accessToken: string): Promise<Subscription | null> {
  return request<Subscription | null>('/subscriptions/me', { accessToken });
}

export async function activateSubscription(
  accessToken: string,
  planCode: string,
): Promise<SubscriptionActivationResponse> {
  return request<SubscriptionActivationResponse>('/subscriptions/activate', {
    method: 'POST',
    accessToken,
    body: JSON.stringify({ plan_code: planCode, provider: 'manual_mvp' }),
  });
}

export async function cancelSubscriptionAtPeriodEnd(accessToken: string): Promise<Subscription> {
  return request<Subscription>('/subscriptions/me/cancel-at-period-end', {
    method: 'POST',
    accessToken,
  });
}

async function request<T = unknown>(
  path: string,
  options: RequestInit & { accessToken?: string } = {},
): Promise<T> {
  const headers = new Headers(options.headers);
  headers.set('Content-Type', 'application/json');
  if (options.accessToken) headers.set('Authorization', `Bearer ${options.accessToken}`);

  const response = await fetch(`${CORE_API_URL}${path}`, {
    ...options,
    headers,
  });

  if (!response.ok) {
    const fallback = `${response.status} ${response.statusText}`;
    let detail = fallback;
    try {
      const body = await response.json();
      detail = typeof body.detail === 'string' ? body.detail : fallback;
    } catch {
      detail = fallback;
    }
    throw new Error(detail);
  }

  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}
